x509/revocation/crl/crl.go

```diff
@@ -0,0 +1,143 @@
+package crl
+
+import (
+	"encoding/asn1"
+	"strconv"
+	"time"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zcrypto/x509/pkix"
+)
+
+// RevocationReasonCode - status codes that explain revocation reason see RFC 5620
+type RevocationReasonCode int
+
+const (
+	Unspecified          RevocationReasonCode = 0
+	KeyCompromise        RevocationReasonCode = 1
+	CACompromise         RevocationReasonCode = 2
+	AffiliationChanged   RevocationReasonCode = 3
+	Superseded           RevocationReasonCode = 4
+	CessationOfOperation RevocationReasonCode = 5
+	CertificateHold      RevocationReasonCode = 6
+	// STATUS CODE 7 IS NOT USED
+	RemoveFromCRL      RevocationReasonCode = 8
+	PrivilegeWithdrawn RevocationReasonCode = 9
+	AACompromise       RevocationReasonCode = 10
+)
+
+func (r RevocationReasonCode) String() string {
+	switch r {
+	case Unspecified:
+		return "Unspecified"
+	case KeyCompromise:
+		return "Key Compromise"
+	case CACompromise:
+		return "CA Compromise"
+	case AffiliationChanged:
+		return "Affiliation Changed"
+	case Superseded:
+		return "Superseded"
+	case CessationOfOperation:
+		return "Cessation Of Operation"
+	case CertificateHold:
+		return "Certificate Hold"
+		// STATUS CODE 7 IS NOT USED
+	case RemoveFromCRL:
+		return "Remove From CRL"
+	case PrivilegeWithdrawn:
+		return "PrivilegeWithdrawn"
+	case AACompromise:
+		return "AA Compromise"
+	default:
+		return "Unknown revocation reason code: " + strconv.Itoa(int(r))
+	}
+}
+
+var (
+	crlNumberExtensionOID        = asn1.ObjectIdentifier{2, 5, 29, 20}
+	revocationReasonExtensionOID = asn1.ObjectIdentifier{2, 5, 29, 21}
+	invalidityDateExtensionOID   = asn1.ObjectIdentifier{2, 5, 29, 24}
+)
+
+type crlNumberExtension struct {
+	ID        asn1.ObjectIdentifier
+	Critical  bool `asn1:"optional"`
+	CRLNumber int
+}
+
+// TODO: handle additional CRL Extensions
+// 2.5.29.21,cRLReason
+// id-ce-cRLReasons OBJECT IDENTIFIER ::= { id-ce 21 }
+// -- reasonCode ::= { CRLReasonCode }
+
+//2.5.29.24,invalidityDate
+// id-ce-invalidityDate OBJECT IDENTIFIER ::= { id-ce 24 }
+// -- InvalidityDate ::=  GeneralizedTime
+
+// RevocationData - information on whether a certificate has been
+// revoked by a specified CRL, and information on the CRL
+type RevocationData struct {
+	CRLSignatureAlgorithm         x509.SignatureAlgorithm
```

Comment:
This goes for all types -- it would be nice if these all had standard `json` fields, and if necessary, custom `MarshalJSON` / 'UnmarshalJSON` methods (say, for `RevocationReasonCode`, it could encode to the string value, but take either the int or string value when decoding)

Reply:
@dadrian how does https://github.com/zmap/zcrypto/pull/142/commits/2715be69b03d0f884f021006341686b441d7aa31 look?